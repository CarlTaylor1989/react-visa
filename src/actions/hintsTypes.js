export const SET_CURRENT_HINT_GROUP = 'SET_CURRENT_HINT_GROUP';
export const SET_HINT_GROUP_INDEX = 'SET_HINT_GROUP_INDEX';
export const SET_HINTS_PAUSED = 'SET_HINTS_PAUSED';
