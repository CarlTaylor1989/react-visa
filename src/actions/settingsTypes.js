export const TOGGLE_AUDIO_STATE = 'TOGGLE_AUDIO_STATE';
export const TOGGLE_TOOLTIPS_STATE = 'TOGGLE_TOOLTIPS_STATE';
export const SET_SETTINGS = 'SET_SETTINGS';
