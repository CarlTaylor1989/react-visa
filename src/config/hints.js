export const HINT_DIAGNOSTIC = 'diagnostic';
export const HINT_NETWORK = 'network';
export const HINT_PRODUCT = 'product';
export const HINT_CLIENT_SELECTION = 'clientselection';
export const HINT_CLIENT_PROPOSAL = 'clientproposal';

export const HINT_TIME = 20000;

export default [
  {
    id: 'h1',
    group: HINT_DIAGNOSTIC
  },
  {
    id: 'h2',
    group: HINT_NETWORK
  },
  {
    id: 'h3',
    group: HINT_NETWORK
  },
  {
    id: 'h4',
    group: HINT_NETWORK
  },
  {
    id: 'h5',
    group: HINT_NETWORK
  },
  {
    id: 'h6',
    group: HINT_NETWORK
  },
  {
    id: 'h7',
    group: HINT_NETWORK
  },
  {
    id: 'h8',
    group: HINT_NETWORK
  },
  {
    id: 'h9',
    group: HINT_PRODUCT
  },
  {
    id: 'h10',
    group: HINT_PRODUCT
  },
  {
    id: 'h11',
    group: HINT_CLIENT_SELECTION
  },
  {
    id: 'h12',
    group: HINT_CLIENT_PROPOSAL
  },
  {
    id: 'h13',
    group: HINT_CLIENT_PROPOSAL
  }
];
