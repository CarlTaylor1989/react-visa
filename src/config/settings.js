/**
 * Setting IDs
 */

export default {
  audio: 'audio-toggle',
  tooltips: 'tooltip-toggle',
  regionBtn: 'region-change',
  tutorial: 'tutorial-toggle',
  tutorialBtn: 'tutorial-reset'
};
