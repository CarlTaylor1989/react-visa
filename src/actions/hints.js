import * as types from './hintsTypes';

export function setCurrentHintGroup(group) {
  return {
    type: types.SET_CURRENT_HINT_GROUP,
    group
  };
}

export function setHintGroupIndex(hints) {
  return {
    type: types.SET_HINT_GROUP_INDEX,
    hints
  };
}

export function setHintsPaused(paused) {
  return {
    type: types.SET_HINTS_PAUSED,
    paused
  };
}
