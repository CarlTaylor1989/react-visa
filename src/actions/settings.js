import * as types from './settingsTypes';

export function toggleAudio(audio) {
  return {
    type: types.TOGGLE_AUDIO_STATE,
    audio
  };
}

export function toggleTooltips(tooltips) {
  return {
    type: types.TOGGLE_TOOLTIPS_STATE,
    tooltips
  };
}

export function setSettings(settings) {
  return {
    type: types.SET_SETTINGS,
    settings
  };
}
