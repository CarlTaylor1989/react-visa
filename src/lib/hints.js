import hints, * as lib from '../config/hints';

/**
 * Filters groups by group into a new array.
 * @param {array} group Array of hints
 * @return {array}
 */
export const getHints = group => hints.filter(hint => hint.group === group);

/**
 * Gets the hint id for a hint based on it's group and index within that group.
 * @param {string} group Group name
 * @param {number} index Current index
 * @return {string | null}
 */
export const getHintId = (group, index) => {
  const hint = getHints(group)[index];
  return hint ? hint.id : null;
};

/**
 * Calculates whether a network hint is accessible.
 * @param {string} hintId Hint identifier
 * @param {object} state Game state
 * @return {boolean}
 */
export const networkHintAccessible = (hintId, state) => {
  let accessible = false;
  switch (hintId) {
    case 'h2':
      accessible = state.completedChallenges.length === 0;
      break;
    case 'h3':
    case 'h4':
    case 'h5':
    case 'h6':
      accessible = state.completedChallenges.length > 0;
      break;
    case 'h7':
      accessible = state.completedProducts.length > 0;
      break;
    case 'h8':
      accessible = state.bonusStreak === 3;
      break;
    default:
      accessible = true;
      break;
  }

  return accessible;
};

/**
 * Calculates whether a product hint is accessible.
 * @param {string} hintId Hint identifier
 * @param {object} state Game state
 * @return {boolean}
 */
export const productHintAccessible = (hintId, state) => {
  let accessible = false;
  switch (hintId) {
    case 'h9':
      accessible = state.completedChallenges.length > 0;
      break;
    case 'h10':
      accessible = state.productPageInactive;
      break;
    default:
      accessible = true;
      break;
  }
  return accessible;
};

export const clientProposalHintAccesible = (hintId, state) => {
  let accessible = false;
  const {
    clients,
    clientShowing,
    displayReqFeedback
  } = state;
  const {
    reqCorrect
  } = clients.find(({ id }) => id === clientShowing);
  switch (hintId) {
    case 'h12':
      accessible = !reqCorrect || (reqCorrect && displayReqFeedback);
      break;
    case 'h13':
      accessible = (reqCorrect && !displayReqFeedback);
      break;
    default:
      accessible = true;
      break;
  }
  return accessible;
};

/**
 * Calculates whether a hint is accessible based on game state. Custom logic for many of the hints.
 * @param {string} group Group name
 * @param {string} hintId Hint identifier
 * @param {object} state Game state
 * @return {boolean}
 */
export const hintAccessible = (group, hintId, state) => {
  let accessible = false;
  if (typeof hintId === 'string') {
    switch (group) {
      case lib.HINT_NETWORK:
        accessible = networkHintAccessible(hintId, state);
        break;
      case lib.HINT_PRODUCT:
        accessible = productHintAccessible(hintId, state);
        break;
      case lib.HINT_CLIENT_PROPOSAL:
        accessible = clientProposalHintAccesible(hintId, state);
        break;
      case lib.HINT_DIAGNOSTIC:
      case lib.HINT_CLIENT_SELECTION:
      default:
        accessible = true;
        break;
    }
  }
  return accessible;
};

/**
 * Recursive function that finds the next valid array index from the
 * array of hints filtered by group, this function takes note of the starting index
 * and returns it if it is reached again as to not end up in a infinite loop of recursive
 * function calls if no hints are currently available for a hint group.
 *
 * @param {string} group Group name
 * @param {number} index Current index
 * @param {object} state Game state
 * @param {number} startingIndex Starting index
 * @return {number}
 */
const getNextHintIndexRecursive = (group, index, state, startingIndex) => {
  const size = getHints(group).length - 1;
  const nextIndex = (index + 1 > size) ? 0 : index + 1;
  const hintId = getHintId(group, nextIndex);
  return (hintAccessible(group, hintId, state) || nextIndex === startingIndex)
    ? nextIndex
    : getNextHintIndexRecursive(group, nextIndex, state, startingIndex);
};

/**
 * Function to find the next valid array index from the
 * array of hints filtered by group.
 * @param {string} group Group name
 * @param {number} index Current index
 * @param {object} state Game state
 * @return {number}
 */
export const getNextHintIndex = (group, index, state) => getNextHintIndexRecursive(
  group,
  index,
  state,
  index
);

/**
 * Gets the current hint index for the current hint group from
 * game state.
 * @param {array} hintsState Array of hints from game state
 * @param {string} group Group name from game state
 * @return {number}
 */
export const getHintIndexFromState = (hintsState, group) => {
  const currentGroup = hintsState.find(hint => hint.group === group);
  return (currentGroup) ? currentGroup.index : null;
};
