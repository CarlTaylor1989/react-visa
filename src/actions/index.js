import { push } from 'connected-react-router';
import * as AchievementActions from './achievements';
import * as AudioActions from './audio';
import * as ClientActions from './client';
import * as DiagnosticActions from './diagnostic';
import * as GenericActions from './generic';
import * as HintsActions from './hints';
import * as MapActions from './map';
import * as PromptActions from './prompts';
import * as ScoreActions from './score';
import * as ScormActions from './scorm';
import * as SettingsActions from './settings';
import * as StreakActions from './streak';
import * as TutorialActions from './tutorial';

const ActionCreators = Object.assign(
  {},
  AchievementActions,
  AudioActions,
  ClientActions,
  DiagnosticActions,
  GenericActions,
  HintsActions,
  MapActions,
  PromptActions,
  ScoreActions,
  ScormActions,
  SettingsActions,
  StreakActions,
  TutorialActions,
  { push }
);

export default ActionCreators;
