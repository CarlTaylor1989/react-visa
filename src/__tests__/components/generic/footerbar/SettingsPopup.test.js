/* global toJson */
import React from 'react'; // eslint-disable-line no-unused-vars
import {
  SettingsPopup,
  mapStateToProps,
  mapDispatchToProps
} from '../../../../components/generic/footerbar/SettingsPopup';
import * as hintActions from '../../../../actions/hints';
import * as settingsActions from '../../../../actions/settings';
import * as hintTypes from '../../../../actions/hintsTypes';
import * as settingsTypes from '../../../../actions/settingsTypes';
import { sendInteractionData } from '../../../../tracking/xapi/XApiAdapter';

jest.mock('../../../../tracking/xapi/XApiAdapter');

let mockProps;

beforeEach(() => {
  mockProps = {
    audio: false,
    changeRegion: jest.fn(),
    closePopup: jest.fn(),
    exitPopup: jest.fn(),
    popupReferrer: 'irrelevant',
    region: '',
    resetDiagnostic: jest.fn(),
    setHintsPaused: jest.fn(),
    showPopup: false,
    toggleAudio: jest.fn(),
    toggleTooltips: jest.fn(),
    tooltips: false
  };
});

test('renders a closed popup', () => {
  const wrapper = shallow(<SettingsPopup {...mockProps} />);
  expect(toJson(wrapper, { noKey: false })).toMatchSnapshot();
});

test('renders an open popup', () => {
  const wrapper = shallow(<SettingsPopup {...mockProps} showPopup />);
  expect(toJson(wrapper, { noKey: false })).toMatchSnapshot();
});

test('renders an open popup with a selected region', () => {
  const wrapper = shallow(<SettingsPopup {...mockProps} region="ac" />);
  expect(toJson(wrapper, { noKey: false })).toMatchSnapshot();
});

describe('method componentDidUpdate()', () => {
  test('popup about to show - pause hints', () => {
    const wrapper = shallow(<SettingsPopup {...mockProps} />);
    wrapper.setProps({ showPopup: true });
    wrapper.update();
    expect(mockProps.setHintsPaused).toHaveBeenCalledWith(true);
  });

  test('popup is still displayed', () => {
    const wrapper = shallow(<SettingsPopup {...mockProps} showPopup />);
    wrapper.setProps({ showPopup: true });
    wrapper.update();
    expect(mockProps.setHintsPaused).not.toHaveBeenCalled();
  });
});

test('method onPopupExit()', () => {
  // setup
  const wrapper = shallow(<SettingsPopup {...mockProps} />);
  wrapper.instance().sendPopupTrackingData = jest.fn();

  // function under test
  wrapper.instance().onPopupExit();

  // expectations
  expect(mockProps.exitPopup).toHaveBeenCalled();
  expect(mockProps.setHintsPaused).toHaveBeenCalledWith(false);
});

test('method onRegionChange()', () => {
  // setup
  const wrapper = shallow(<SettingsPopup {...mockProps} />);
  wrapper.instance().sendPopupTrackingData = jest.fn();

  // function under test
  wrapper.instance().onRegionChange();

  // expectations
  expect(mockProps.changeRegion).toHaveBeenCalled();
  expect(wrapper.instance().sendPopupTrackingData).toHaveBeenCalled();
});

test('method sendPopupTrackingData', () => {
  const wrapper = shallow(<SettingsPopup {...mockProps} />);
  wrapper.instance().sendPopupTrackingData();
  expect(sendInteractionData).toHaveBeenCalledWith(expect.objectContaining({
    type: 'experienced',
    activity: expect.objectContaining({
      id: 'settings',
      type: 'slide',
      title: expect.any(String),
      description: expect.any(String)
    }),
    response: expect.objectContaining({
      timeAvailable: null,
      timeResponse: expect.any(Date)
    }),
    parent: '',
    referrer: expect.any(String)
  }));
});

test('should mapStateToProps', () => {
  const state = {
    diagnosticData: {
      region: 'eu'
    },
    genericData: {
      popupReferrer: 'irrelevant'
    },
    settingsData: {
      tooltips: true,
      audio: true
    }
  };
  const expected = {
    ...state.settingsData,
    ...state.diagnosticData,
    ...state.genericData
  };
  const props = mapStateToProps(state);
  expect(props).toEqual({
    ...expected
  });
});

test('calls toggleTooltips', () => {
  const dispatch = jest.fn();
  jest.spyOn(settingsActions, 'toggleTooltips').mockReturnValue('toggleTooltips()');
  const props = mapDispatchToProps(dispatch);
  props.toggleTooltips();
  expect(dispatch).toHaveBeenCalledWith({
    type: settingsTypes.TOGGLE_TOOLTIPS_STATE
  });
});

test('calls toggleAudio', () => {
  const dispatch = jest.fn();
  jest.spyOn(settingsActions, 'toggleAudio').mockReturnValue('toggleAudio()');
  const props = mapDispatchToProps(dispatch);
  props.toggleAudio();
  expect(dispatch).toHaveBeenCalledWith({
    type: settingsTypes.TOGGLE_AUDIO_STATE
  });
});

test('calls setHintsPaused', () => {
  // setup
  const dispatch = jest.fn();
  jest.spyOn(
    hintActions,
    'setHintsPaused'
  ).mockReturnValue('setHintsPaused()');
  const props = mapDispatchToProps(dispatch);

  // function under test
  props.setHintsPaused();

  // expectations
  expect(dispatch).toHaveBeenCalledWith({
    type: hintTypes.SET_HINTS_PAUSED
  });
});
