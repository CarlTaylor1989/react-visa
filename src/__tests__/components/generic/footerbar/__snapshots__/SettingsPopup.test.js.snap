// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a closed popup 1`] = `
<Modal
  animationDuration={500}
  blockScroll={true}
  center={true}
  classNames={
    Object {
      "modal": "settingsPopup",
      "overlay": "settingsPopupOverlay",
    }
  }
  classes={
    Object {
      "closeButton": "styles_closeButton__20ID4",
      "closeIcon": "styles_closeIcon__1QwbI",
      "modal": "styles_modal__gNwvD",
      "modalCenter": "styles_modalCenter__L9F2w",
      "overlay": "styles_overlay__CLSq-",
      "transitionEnter": "styles_transitionEnter__3j_-a",
      "transitionEnterActive": "styles_transitionEnterActive___eQs7",
      "transitionExit": "styles_transitionExit__1KmEf",
      "transitionExitActive": "styles_transitionExitActive__1nQXw",
    }
  }
  closeIconSize={28}
  closeIconSvgPath={
    <path
      d="M28.5 9.62L26.38 7.5 18 15.88 9.62 7.5 7.5 9.62 15.88 18 7.5 26.38l2.12 2.12L18 20.12l8.38 8.38 2.12-2.12L20.12 18z"
    />
  }
  closeOnEsc={false}
  closeOnOverlayClick={false}
  container={null}
  focusTrapOptions={Object {}}
  focusTrapped={true}
  onClose={[MockFunction]}
  onExited={[Function]}
  open={false}
  role="dialog"
  showCloseIcon={false}
  styles={Object {}}
>
  <button
    className="exitPopup"
    onClick={[MockFunction]}
    type="button"
  >
    <Translate
      tag="span"
      value="generic.popupCloseBtn"
    />
  </button>
  <div
    className="popupWrapper"
  >
    <div
      className="popupTitle"
    >
      <Translate
        tag="span"
        value="generic.settings.popupTitle"
      />
    </div>
    <div
      className="settings"
    >
      <SettingRow
        buttonLabel=""
        checked={false}
        extraText=""
        handleChange={[MockFunction]}
        handleClick={null}
        hasButton={false}
        hasToggle={true}
        name="audio"
      />
      <SettingRow
        buttonLabel="change"
        checked={false}
        extraText=""
        handleChange={null}
        handleClick={[Function]}
        hasButton={true}
        hasToggle={false}
        name="region"
      />
      <SettingRow
        buttonLabel=""
        checked={false}
        extraText=""
        handleChange={[MockFunction]}
        handleClick={null}
        hasButton={false}
        hasToggle={true}
        name="tooltips"
      />
    </div>
  </div>
</Modal>
`;

exports[`renders an open popup 1`] = `
<Modal
  animationDuration={500}
  blockScroll={true}
  center={true}
  classNames={
    Object {
      "modal": "settingsPopup",
      "overlay": "settingsPopupOverlay",
    }
  }
  classes={
    Object {
      "closeButton": "styles_closeButton__20ID4",
      "closeIcon": "styles_closeIcon__1QwbI",
      "modal": "styles_modal__gNwvD",
      "modalCenter": "styles_modalCenter__L9F2w",
      "overlay": "styles_overlay__CLSq-",
      "transitionEnter": "styles_transitionEnter__3j_-a",
      "transitionEnterActive": "styles_transitionEnterActive___eQs7",
      "transitionExit": "styles_transitionExit__1KmEf",
      "transitionExitActive": "styles_transitionExitActive__1nQXw",
    }
  }
  closeIconSize={28}
  closeIconSvgPath={
    <path
      d="M28.5 9.62L26.38 7.5 18 15.88 9.62 7.5 7.5 9.62 15.88 18 7.5 26.38l2.12 2.12L18 20.12l8.38 8.38 2.12-2.12L20.12 18z"
    />
  }
  closeOnEsc={false}
  closeOnOverlayClick={false}
  container={null}
  focusTrapOptions={Object {}}
  focusTrapped={true}
  onClose={[MockFunction]}
  onExited={[Function]}
  open={true}
  role="dialog"
  showCloseIcon={false}
  styles={Object {}}
>
  <button
    className="exitPopup"
    onClick={[MockFunction]}
    type="button"
  >
    <Translate
      tag="span"
      value="generic.popupCloseBtn"
    />
  </button>
  <div
    className="popupWrapper"
  >
    <div
      className="popupTitle"
    >
      <Translate
        tag="span"
        value="generic.settings.popupTitle"
      />
    </div>
    <div
      className="settings"
    >
      <SettingRow
        buttonLabel=""
        checked={false}
        extraText=""
        handleChange={[MockFunction]}
        handleClick={null}
        hasButton={false}
        hasToggle={true}
        name="audio"
      />
      <SettingRow
        buttonLabel="change"
        checked={false}
        extraText=""
        handleChange={null}
        handleClick={[Function]}
        hasButton={true}
        hasToggle={false}
        name="region"
      />
      <SettingRow
        buttonLabel=""
        checked={false}
        extraText=""
        handleChange={[MockFunction]}
        handleClick={null}
        hasButton={false}
        hasToggle={true}
        name="tooltips"
      />
    </div>
  </div>
</Modal>
`;

exports[`renders an open popup with a selected region 1`] = `
<Modal
  animationDuration={500}
  blockScroll={true}
  center={true}
  classNames={
    Object {
      "modal": "settingsPopup",
      "overlay": "settingsPopupOverlay",
    }
  }
  classes={
    Object {
      "closeButton": "styles_closeButton__20ID4",
      "closeIcon": "styles_closeIcon__1QwbI",
      "modal": "styles_modal__gNwvD",
      "modalCenter": "styles_modalCenter__L9F2w",
      "overlay": "styles_overlay__CLSq-",
      "transitionEnter": "styles_transitionEnter__3j_-a",
      "transitionEnterActive": "styles_transitionEnterActive___eQs7",
      "transitionExit": "styles_transitionExit__1KmEf",
      "transitionExitActive": "styles_transitionExitActive__1nQXw",
    }
  }
  closeIconSize={28}
  closeIconSvgPath={
    <path
      d="M28.5 9.62L26.38 7.5 18 15.88 9.62 7.5 7.5 9.62 15.88 18 7.5 26.38l2.12 2.12L18 20.12l8.38 8.38 2.12-2.12L20.12 18z"
    />
  }
  closeOnEsc={false}
  closeOnOverlayClick={false}
  container={null}
  focusTrapOptions={Object {}}
  focusTrapped={true}
  onClose={[MockFunction]}
  onExited={[Function]}
  open={false}
  role="dialog"
  showCloseIcon={false}
  styles={Object {}}
>
  <button
    className="exitPopup"
    onClick={[MockFunction]}
    type="button"
  >
    <Translate
      tag="span"
      value="generic.popupCloseBtn"
    />
  </button>
  <div
    className="popupWrapper"
  >
    <div
      className="popupTitle"
    >
      <Translate
        tag="span"
        value="generic.settings.popupTitle"
      />
    </div>
    <div
      className="settings"
    >
      <SettingRow
        buttonLabel=""
        checked={false}
        extraText=""
        handleChange={[MockFunction]}
        handleClick={null}
        hasButton={false}
        hasToggle={true}
        name="audio"
      />
      <SettingRow
        buttonLabel="change"
        checked={false}
        extraText="code"
        handleChange={null}
        handleClick={[Function]}
        hasButton={true}
        hasToggle={false}
        name="region"
      />
      <SettingRow
        buttonLabel=""
        checked={false}
        extraText=""
        handleChange={[MockFunction]}
        handleClick={null}
        hasButton={false}
        hasToggle={true}
        name="tooltips"
      />
    </div>
  </div>
</Modal>
`;
